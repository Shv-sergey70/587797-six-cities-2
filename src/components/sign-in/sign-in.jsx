import React from 'react';
import PropTypes from 'prop-types';
import Operation from "../../operation";
import {connect} from "react-redux";
import history from "../../history";
import {Redirect} from 'react-router-dom';
import {Route} from "../../const/routes";
import PageHeader from "../page-header/page-header";
import Selectors from '../../selector';

class SignIn extends React.PureComponent {
  constructor(props) {
    super(props);

    this._onSignInButtonClick = this._onSignInButtonClick.bind(this);
  }

  render() {
    const {
      email,
      password,
      onInputKeydown,
      isAuth
    } = this.props;

    if (isAuth) {
      return <Redirect to={Route.MAIN} />;
    }

    return <div className="page page--gray page--login">
      <PageHeader/>

      <main className="page__main page__main--login">
        <div className="page__login-container container">
          <section className="login">
            <h1 className="login__title">Sign in</h1>
            <form className="login__form form" action="#" method="post">
              <div className="login__input-wrapper form__input-wrapper">
                <label className="visually-hidden">E-mail</label>
                <input className="login__input form__input" type="email" name="email" placeholder="Email" required="" defaultValue={email} onChange={onInputKeydown}/>
              </div>
              <div className="login__input-wrapper form__input-wrapper">
                <label className="visually-hidden">Password</label>
                <input className="login__input form__input" type="password" name="password" placeholder="Password" required="" defaultValue={password} onChange={onInputKeydown}/>
              </div>
              <button className="login__submit form__submit button" type="submit" onClick={this._onSignInButtonClick}>Sign in</button>
            </form>
          </section>
          <section className="locations locations--login locations--current">
            <div className="locations__item">
              <a className="locations__item-link" href="#">
                <span>Amsterdam</span>
              </a>
            </div>
          </section>
        </div>
      </main>
    </div>;
  }

  _onSignInButtonClick(evt) {
    evt.preventDefault();

    const {
      email,
      password,
      authorize
    } = this.props;

    if (email.length === 0) {
      throw new Error(`Email is empty`);
    } else if (password.length === 0) {
      throw new Error(`Password is empty`);
    }

    authorize(email, password);
    history.push(Route.MAIN);
  }
}

SignIn.propTypes = {
  email: PropTypes.string,
  password: PropTypes.string,
  onInputKeydown: PropTypes.func.isRequired,
  authorize: PropTypes.func.isRequired,
  isAuth: PropTypes.bool.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  isAuth: !!Selectors.getUserEmail(state)
});

const mapDispatchToProps = (dispatch) => ({
  authorize: (email, password) => dispatch(Operation.authorize(email, password))
});

export {SignIn};
export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
