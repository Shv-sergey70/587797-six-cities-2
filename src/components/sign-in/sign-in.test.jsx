import React from 'react';
import renderer from 'react-test-renderer';
import {SignIn} from "./sign-in";

jest.mock(`../page-header/page-header.jsx`, () => `PageHeader`);

it(`Sign-in page renders correctly`, () => {
  const tree = renderer.create(<SignIn
    email={`test@test.ru`}
    password={`my_password`}
    onInputKeydown={jest.fn()}
    authorize={jest.fn()}
    isAuth={false}
  />).toJSON();

  expect(tree).toMatchSnapshot();
});
