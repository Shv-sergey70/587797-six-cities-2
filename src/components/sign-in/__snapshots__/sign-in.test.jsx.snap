// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign-in page renders correctly 1`] = `
<div
  className="page page--gray page--login"
>
  <PageHeader />
  <main
    className="page__main page__main--login"
  >
    <div
      className="page__login-container container"
    >
      <section
        className="login"
      >
        <h1
          className="login__title"
        >
          Sign in
        </h1>
        <form
          action="#"
          className="login__form form"
          method="post"
        >
          <div
            className="login__input-wrapper form__input-wrapper"
          >
            <label
              className="visually-hidden"
            >
              E-mail
            </label>
            <input
              className="login__input form__input"
              defaultValue="test@test.ru"
              name="email"
              onChange={[MockFunction]}
              placeholder="Email"
              required=""
              type="email"
            />
          </div>
          <div
            className="login__input-wrapper form__input-wrapper"
          >
            <label
              className="visually-hidden"
            >
              Password
            </label>
            <input
              className="login__input form__input"
              defaultValue="my_password"
              name="password"
              onChange={[MockFunction]}
              placeholder="Password"
              required=""
              type="password"
            />
          </div>
          <button
            className="login__submit form__submit button"
            onClick={[Function]}
            type="submit"
          >
            Sign in
          </button>
        </form>
      </section>
      <section
        className="locations locations--login locations--current"
      >
        <div
          className="locations__item"
        >
          <a
            className="locations__item-link"
            href="#"
          >
            <span>
              Amsterdam
            </span>
          </a>
        </div>
      </section>
    </div>
  </main>
</div>
`;
